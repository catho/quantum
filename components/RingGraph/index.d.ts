
// Generated with scripts/create-component.js
import { FC } from 'react';

export interface RingGraphProps {
  percentage: number;
  round: boolean;
  skin: 'primary' | 'secondary' | 'success' | 'warning' | 'error' | 'neutral';
  theme?: { baseFontSize?: number; colors?: object };
}

declare const RingGraph: FC<RingGraphProps>;
export default RingGraph;
    