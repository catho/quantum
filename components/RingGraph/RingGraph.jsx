// Generated with scripts/create-component.js
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import styled, { css } from 'styled-components';
import { colors, baseFontSize as defaultBaseFontSize } from '../shared/theme';
import { shadow } from '../shared';
import { percentToDegrees } from '../Stepper/helper';

const Wrapper = styled.div`
  display: inline-flex;
  align-items: center;
`;

const Circle = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  min-width: 24px;
  height: 24px;
  min-height: 24px;

  ${({
    degrees: { 0: reference, 1: position },
    skin,
    theme: {
      colors: {
        neutral: { 300: neutralColor },
        [skin]: { 900: color },
      },
    },
  }) =>
    reference > 0
      ? css`
          background-image: linear-gradient(
              ${reference}deg,
              ${neutralColor} 50%,
              transparent 50%
            ),
            linear-gradient(${position}deg, ${neutralColor} 50%, ${color} 50%);
        `
      : css`
          background-image: linear-gradient(
              ${reference}deg,
              ${color} 50%,
              transparent 50%
            ),
            linear-gradient(${position}deg, ${color} 50%, ${neutralColor} 50%);
        `}
  border-radius: 50%;
  ${shadow(2)}
`;
const CircleOverlay = styled.div`
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: ${({
    theme: {
      colors: {
        neutral: { 0: neutralColor },
      },
    },
  }) => neutralColor};
`;

const ProgressWrapper = styled.span`
  margin-left: 12px;
  ${({
    theme: {
      baseFontSize,
      colors: {
        primary: { 900: primaryColor },
      },
    },
  }) => `
    color: ${primaryColor};
    font-size: ${baseFontSize}px;
  `}
`;
const formatInPercentage = (number) =>
  number.toLocaleString('pt-br', {
    style: 'percent',
    maximumFractionDigits: 2,
  });

const RingGraph = ({ percentage, round, skin, theme }) => {
  const [compatibilityPercentage, setCompatibilityPercentage] = useState(
    formatInPercentage(0),
  );

  useEffect(() => {
    if (percentage >= 0 && percentage <= 100) {
      let percentageValue = percentage;
      if (round) {
        percentageValue = Math.round(percentageValue);
      }
      const formattedPercentageValue = formatInPercentage(
        percentageValue / 100,
      );
      setCompatibilityPercentage(formattedPercentageValue);
    }
  }, [percentage]);

  return (
    <Wrapper>
      <Circle degrees={percentToDegrees(percentage)} theme={theme} skin={skin}>
        <CircleOverlay theme={theme} />
      </Circle>
      <ProgressWrapper
        theme={theme}
      >{`${compatibilityPercentage} compat√≠vel`}</ProgressWrapper>
    </Wrapper>
  );
};

RingGraph.propTypes = {
  percentage: PropTypes.number,
  round: PropTypes.bool,
  skin: PropTypes.oneOf([
    'primary',
    'secondary',
    'success',
    'warning',
    'error',
    'neutral',
  ]),
  theme: PropTypes.shape({
    colors: PropTypes.object,
    baseFontSize: PropTypes.number,
  }),
};

RingGraph.defaultProps = {
  percentage: 0,
  round: true,
  skin: 'primary',
  theme: {
    colors,
    baseFontSize: defaultBaseFontSize,
  },
};

export default RingGraph;
