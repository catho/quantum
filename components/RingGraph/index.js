// Generated with scripts/create-component.js
import RingGraph from './RingGraph';

export default RingGraph;
