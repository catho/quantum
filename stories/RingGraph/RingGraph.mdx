{/* Generated with scripts/create-componnent.js */}
import { Meta, Canvas, Story, ArgsTable } from '@storybook/addon-docs';
import  RingGraph  from '../../components/RingGraph';

import { Header } from '../shared';

import { 
  Default
} from './RingGraph.stories.jsx';

<Meta title="RingGraph" component={RingGraph}  />

<Header title="RingGraph">
  The RingGraph component is used to update the user on the percentage of compatibility with a job vacancy
</Header>

- [Importing and usage](#importing)
- [API](#api)

## <a id="importing"></a>Importing and usage

```js
import { RingGraph } from '@catho/quantum';
```

<Canvas>
  <Story story={Default} />
</Canvas>

## <a id="api"></a>API

<ArgsTable of={RingGraph} />
    