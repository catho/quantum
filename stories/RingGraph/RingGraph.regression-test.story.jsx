// Generated with scripts/create-componnent.js

import RingGraph from '../../components/RingGraph';

export default {
  title: 'RingGraph',
  component: RingGraph,
};

const Template = (args) => <RingGraph {...args} />;

export const Default = Template.bind({});
Default.args = {
  percentage: 0,
};

export const WithoutRouded = Template.bind({});
WithoutRouded.args = {
  percentage: 55.55,
  round: false,
};

export const WithSkin = Template.bind({});
WithSkin.args = {
  percentage: 50,
  skin: 'secondary',
};
