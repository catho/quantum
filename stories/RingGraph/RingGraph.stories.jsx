// Generated with scripts/create-componnent.js

import RingGraph from '../../components/RingGraph';

export default {
  title: 'RingGraph',
  component: RingGraph,
};

const Template = (args) => <RingGraph {...args} />;

export const Default = Template.bind({});
Default.args = {
  percentage: 25,
};
