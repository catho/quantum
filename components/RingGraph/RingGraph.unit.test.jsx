// Generated with scripts/create-component.js
import { render, screen } from '@testing-library/react';

import RingGraph from './RingGraph';

describe('<RingGraph />', () => {
  it('should match the snapshots', () => {
    expect(render(<RingGraph percentage={0} />).asFragment()).toMatchSnapshot();

    expect(
      render(<RingGraph percentage={25} />).asFragment(),
    ).toMatchSnapshot();

    expect(
      render(<RingGraph percentage={50} />).asFragment(),
    ).toMatchSnapshot();

    expect(
      render(<RingGraph percentage={75} />).asFragment(),
    ).toMatchSnapshot();

    expect(
      render(<RingGraph percentage={100} />).asFragment(),
    ).toMatchSnapshot();

    expect(
      render(<RingGraph percentage={100} skin="secondary" />).asFragment(),
    ).toMatchSnapshot();
  });

  it('should round percentage when value is float', () => {
    render(<RingGraph percentage={50.5} />);

    expect(screen.getByText(/51% compatível/)).toBeInTheDocument();
  });

  it('should display percentage 0 when prop "percentage" is less than 100', () => {
    render(<RingGraph percentage={-1} />);

    expect(screen.getByText(/0% compatível/)).toBeInTheDocument();
  });

  it('should display percentage 0 when prop "percentage" is greater than 100', () => {
    render(<RingGraph percentage={110} />);

    expect(screen.getByText(/0% compatível/)).toBeInTheDocument();
  });

  it('should round percentage when value is float', () => {
    render(<RingGraph percentage={50.5} />);

    expect(screen.getByText(/51% compatível/)).toBeInTheDocument();
  });

  it('should display actual percentage value when not an integer and when value of prop "round" is false', () => {
    render(<RingGraph percentage={50.5} round={false} />);

    expect(screen.getByText(/50,5% compatível/)).toBeInTheDocument();
  });

  it('must show maximum two digits after comma when percentage is float and has more than 2 digits after dot and prop "round" is false', () => {
    render(<RingGraph percentage={50.555} round={false} />);

    expect(screen.getByText(/50,55% compatível/)).toBeInTheDocument();
  });
});
