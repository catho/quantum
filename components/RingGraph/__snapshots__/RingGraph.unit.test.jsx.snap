// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RingGraph /> should match the snapshots 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  min-width: 24px;
  height: 24px;
  min-height: 24px;
  background-image: linear-gradient(90deg,#e0e0e0 50%,transparent 50%),linear-gradient(-90deg,#e0e0e0 50%,#002F7B 50%);
  border-radius: 50%;
  box-shadow: 0px 3px 1px -2px rgba(66,66,66,0.2),0px 2px 2px 0px rgba(66,66,66,0.14),0px 1px 5px 0px rgba(66,66,66,0.12);
}

.c2 {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: #ffffff;
}

.c3 {
  margin-left: 12px;
  color: #002F7B;
  font-size: 16px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c3"
    >
      0% compatível
    </div>
  </div>
</DocumentFragment>
`;

exports[`<RingGraph /> should match the snapshots 2`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  min-width: 24px;
  height: 24px;
  min-height: 24px;
  background-image: linear-gradient(90deg,#e0e0e0 50%,transparent 50%),linear-gradient(0deg,#e0e0e0 50%,#002F7B 50%);
  border-radius: 50%;
  box-shadow: 0px 3px 1px -2px rgba(66,66,66,0.2),0px 2px 2px 0px rgba(66,66,66,0.14),0px 1px 5px 0px rgba(66,66,66,0.12);
}

.c2 {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: #ffffff;
}

.c3 {
  margin-left: 12px;
  color: #002F7B;
  font-size: 16px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c3"
    >
      25% compatível
    </div>
  </div>
</DocumentFragment>
`;

exports[`<RingGraph /> should match the snapshots 3`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  min-width: 24px;
  height: 24px;
  min-height: 24px;
  background-image: linear-gradient(90deg,#e0e0e0 50%,transparent 50%),linear-gradient(90deg,#e0e0e0 50%,#002F7B 50%);
  border-radius: 50%;
  box-shadow: 0px 3px 1px -2px rgba(66,66,66,0.2),0px 2px 2px 0px rgba(66,66,66,0.14),0px 1px 5px 0px rgba(66,66,66,0.12);
}

.c2 {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: #ffffff;
}

.c3 {
  margin-left: 12px;
  color: #002F7B;
  font-size: 16px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c3"
    >
      50% compatível
    </div>
  </div>
</DocumentFragment>
`;

exports[`<RingGraph /> should match the snapshots 4`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  min-width: 24px;
  height: 24px;
  min-height: 24px;
  background-image: linear-gradient(-90deg,#002F7B 50%,transparent 50%),linear-gradient(0deg,#002F7B 50%,#e0e0e0 50%);
  border-radius: 50%;
  box-shadow: 0px 3px 1px -2px rgba(66,66,66,0.2),0px 2px 2px 0px rgba(66,66,66,0.14),0px 1px 5px 0px rgba(66,66,66,0.12);
}

.c2 {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: #ffffff;
}

.c3 {
  margin-left: 12px;
  color: #002F7B;
  font-size: 16px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c3"
    >
      75% compatível
    </div>
  </div>
</DocumentFragment>
`;

exports[`<RingGraph /> should match the snapshots 5`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  min-width: 24px;
  height: 24px;
  min-height: 24px;
  background-image: linear-gradient(-90deg,#002F7B 50%,transparent 50%),linear-gradient(90deg,#002F7B 50%,#e0e0e0 50%);
  border-radius: 50%;
  box-shadow: 0px 3px 1px -2px rgba(66,66,66,0.2),0px 2px 2px 0px rgba(66,66,66,0.14),0px 1px 5px 0px rgba(66,66,66,0.12);
}

.c2 {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: #ffffff;
}

.c3 {
  margin-left: 12px;
  color: #002F7B;
  font-size: 16px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c3"
    >
      100% compatível
    </div>
  </div>
</DocumentFragment>
`;

exports[`<RingGraph /> should match the snapshots 6`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  min-width: 24px;
  height: 24px;
  min-height: 24px;
  background-image: linear-gradient(-90deg,#9E003F 50%,transparent 50%),linear-gradient(90deg,#9E003F 50%,#e0e0e0 50%);
  border-radius: 50%;
  box-shadow: 0px 3px 1px -2px rgba(66,66,66,0.2),0px 2px 2px 0px rgba(66,66,66,0.14),0px 1px 5px 0px rgba(66,66,66,0.12);
}

.c2 {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: #ffffff;
}

.c3 {
  margin-left: 12px;
  color: #002F7B;
  font-size: 16px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c3"
    >
      100% compatível
    </div>
  </div>
</DocumentFragment>
`;
